man/systemd-run-generator.xml

```diff
@@ -0,0 +1,76 @@
+<?xml version="1.0"?>
+<!--*-nxml-*-->
+<!DOCTYPE refentry PUBLIC "-//OASIS//DTD DocBook XML V4.2//EN" "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
+<!--
+  SPDX-License-Identifier: LGPL-2.1+
+-->
+<refentry id="systemd-run-generator">
+
+  <refentryinfo>
+    <title>systemd-run-generator</title>
+    <productname>systemd</productname>
+  </refentryinfo>
+
+  <refmeta>
+    <refentrytitle>systemd-run-generator</refentrytitle>
+    <manvolnum>8</manvolnum>
+  </refmeta>
+
+  <refnamediv>
+    <refname>systemd-run-generator</refname>
+    <refpurpose>Generator for invoking commands specified on the kernel command line as system service</refpurpose>
+  </refnamediv>
+
+  <refsynopsisdiv>
+    <para><filename>/usr/lib/systemd/system-generators/systemd-run-generator</filename></para>
+  </refsynopsisdiv>
+
+  <refsect1>
+    <title>Description</title>
+
+    <para><filename>systemd-run-generator</filename> is a generator
+    that reads the kernel command line and understands three
+    options:</para>
+
+    <para>If the <option>systemd.run=</option> option is specified and followed by a command line a unit named
+    <filename>kernel-command-line.service</filename> is generated for it and pulled into the initial boot-up
+    transaction. The service has <varname>Type=oneshot</varname> set, and has <varname>SuccessAction=exit</varname> and
```

Comment:
This is not obvious from this description, but looking at the code, the generated service is simply added as a dependency of `default.target`. I don't think this is as useful as could be, because all the stuff that is normally started is started here too.

There are two orthogonal issues:
1. whether we add a unit to the default target or whether we replace the default target (`systemd.unit=` does the latter, we don't have a mechanism for the former) ("role").
2. whether the unit is defined using `systemd-run` syntax (which this PR does) or whether an existing unit is necessary (what systemd.unit= requires) ("syntax").

So for example, to run `adduser testuser` we do not want to start default.target. Instead, we want to pull in the dependencies as specified by DefaultDependencies and nothing else. But on the other hand, if I want to start a container with 'pcp.service' additionally active to monitor resource usage or a custom `pcp` command with some switches, I just want to add the service to the existing transaction.

I think that it'd be good if the syntax was flexible enough, forward-looking enough, to support all four cases: adding a unit to the default target, replacing the default target with a unit (already supported), adding a run command to the default target (this PR), replacing the default target with a run command. In other words, to determine the "role" independently of "syntax", even if only parts of the functionality are implemented now. Right now the semantics feel quite ad-hoc.

So I'd propose the following changes:
a. make `systemd.run=` unit replace default target (changing this PR)
b. add `systemd.unit_success_action=` and `systemd.unit_failure_action=` (some future PR)
c. add some switch to add units to the default target instead of replacing it (some future PR, but it'd be good to decide the interface now, so we don't paint ourselves into a corner). This would apply to be `.unit=` and `.run=` both.

Some idea for syntax for (c):
- `systemd.wants-unit=` and `systemd.requires-unit=` and `systemd.wants-run=` and and `systemd.requires-run=`
- `systemd.add-unit=` and `systemd.add-run=`, and if `-` is used as the first character of the command, it means .wants, and .requires otherwise.

Then there's the question what happens when `systemd.unit=` and `systemd.run=` are both specified. Right now `unit=` effectively trumps the other one. If my suggested change (a) above is implemented, this will be moot. If (c) is ever implemented, the generator should be smarter, and if `systemd.unit=` is also present, it should add the generated service to that target instead.

Reply:
i agree, will rework